### PR TITLE
Tiny nft username issue 767

### LINKED ISSUE #737: Ability to transfer a name to another owner
**Is your feature request related to a problem? Please describe.**
When a name is registered it belongs to one address and cannot be transferred.

**Describe the solution you'd like**
As a user I want to be able to transfer my name to another address.

**Describe alternatives you've considered**
Not applicable

**Additional context**
For Weave 0.17 release

### PR DESCRIPTION
**Pull request by husio**
close #737 

This is a first iteration and proof of concept for the "tiny username"™ implementation. The idea was described in #767 

_:warning: There are no tests, `x/username` was added without removing `nft/username` and some rules (ie validation) are objectively selected. If this approach is accepted, those shortcuts must be fixed._

The most important file is the protobuf declaration.

Tested only using local bnsd instance.

```sh
$ bnscli register-username -h

Create a transaction for registering a username.

  -addr value
        Hex encoded blochain address on this network.
  -bc string
        Blockchain network ID.
  -name string
        Name part of the username. For example 'alice'
  -ns string
        Namespace (domain) part of the username. For example 'iov' (default "iov")

$ bnscli register-username -name bob -ns fairycoin -bc secretbc -addr 46722031342e204a756e2031303a35363a3337204345535420323031390a \
        | bnscli with-fee -amount "1 CASH" \
        | bnscli sign \
        | bnscli submit                           

$ bnscli resolve-username -name bob -ns fairycoin                                                                                                    
{
        "metadata": {
                "schema": 1
        },
        "target": {
                "blockchain_id": "secretbc",
                "address": "RnIgMTQuIEp1biAxMDo1NjozNyBDRVNUIDIwMTkK"
        },
        "owner": "63BDD1C037BF131A9A374399AE6B3781C75C1563"
}
```

### CODE REVIEW

**Comment by ruseinov** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: spec/gogo/x/username/codec.proto [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  Location target = 2 [(gogoproto.nullable) = false];
+  // Owner is a weave.Address that controls this token. Only the owner can
+  // modify a username token.
+  bytes owner = 3 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+}
+
+// Location represents a blochain address. This structure clubs together
+// blokchain ID together with an address on that network. It is used to point
+// to an address on any blockchain network.
+message Location {
```
Comment (alpe) [inline spec/gogo/x/username/codec.proto]: pp: naming.... I found this more confusing than `ChainAddress`

File: spec/gogo/x/username/codec.proto [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+// that the name is unique. Username is a combination of a name and a domain.
+// The format is <name>*<domain>
+//
+// Each token points to a blockchain and an address on that blockchain. Both
+// blockchain ID and address are an arbitrary string as we do not want to limit
+// ourselves to certain patterns.
+message Token {
+  weave.Metadata metadata = 1;
+  // Target specifies where this username token points to.
+  Location target = 2 [(gogoproto.nullable) = false];
```
Comment (alpe) [inline spec/gogo/x/username/codec.proto]: req: a single username can point to multiple chain addresses

File: x/username/codec.proto [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  // Username is the unique name of the token, for example alice*iov
+  string username = 2 [(gogoproto.casttype) = "Username"];
+  // Target is a blockchain address location that this token should point to.
+  Location target = 3 [(gogoproto.nullable) = false];
+}
+
+// ChangeTokenOwnerMsg is a request to transfer an ownership of a token. The
+// mesage must be signed by the current token owner. Acceptance of the new
+// owner is not required in order to succeed.
+message ChangeTokenOwnerMsg {
```
Comment (alpe) [inline x/username/codec.proto]: pp: how about TransferTokenMsg?

File: x/username/codec.proto [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  weave.Metadata metadata = 1;
+  // Username is the unique name of the token, for example alice*iov
+  string username = 2 [(gogoproto.casttype) = "Username"];
+  // Owner is a weave address that will owns this token after the change.
+  bytes new_owner = 3 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+}
+
+// ChangeTokenTargetMsg is a request to change the address that this token
+// points to. Only the owner of a token can request this operation.
+message ChangeTokenTargetMsg {
```
Comment (alpe) [inline x/username/codec.proto]: pp: UpdateChainAddress? Do you want to support multiple at once?

File: cmd/bnscli/cmd_username.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+	raw, err := json.MarshalIndent(token, "", "\t")
+	if err != nil {
+		return fmt.Errorf("cannot json serialize token information: %s", err)
+	}
+	_, err = output.Write(raw)
+	return err
+}
+
+func fetchUsernameToken(serverURL string, uname username.Username) (*username.Token, error) {
```
Comment (ethanfrey) [inline cmd/bnscli/cmd_username.go]: This is a lot of detail work. We should add better helpers for queries.

This is not a comment on the pr, but on the cli in general..

Comment (ethanfrey) [reply cmd/bnscli/cmd_username.go]: And I think it is my open issue since 2 months

File: spec/gogo/x/username/codec.proto (hunk 1 above)
Comment (ruseinov) [reply spec/gogo/x/username/codec.proto]: I second that - looks confusing to me, as I start thinking about physical location of things. 
I would either keep `ChainAddress` which is not fantastic, or introduce an `Address` entity, which makes sense in terms of NFT. In the end `ChainAddress` is more self-explanatory. 

It could also be `ExternalAddress` or something similar.

File: x/username/codec.proto [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+// The format is <name>*<domain>
+//
+// Each token points to a blockchain and an address on that blockchain. Both
+// blockchain ID and address are an arbitrary string as we do not want to limit
+// ourselves to certain patterns.
+message Token {
+  weave.Metadata metadata = 1;
+  // Target specifies where this username token points to.
+  Location target = 2 [(gogoproto.nullable) = false];
+  // Owner is a weave.Address that controls this token. Only the owner can
```
Comment (ethanfrey) [inline x/username/codec.proto]: This is fine for the demo.
For the actual minimal usecase to support our framework, we need a list of chain, address pairs. The wallet on creation registers addresses on 3 or 4 chains by default

File: x/username/codec.proto [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  // modify a username token.
+  bytes owner = 3 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+}
+
+// Location represents a blochain address. This structure clubs together
+// blokchain ID together with an address on that network. It is used to point
+// to an address on any blockchain network.
+message Location {
```
Comment (ethanfrey) [inline x/username/codec.proto]: Nice name. Just include an array (which makes a bunch of boilerplate)

File: x/username/handlers.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+	registerTokenCost     = 0
+	changeTokenOwnerCost  = 0
+	changeTokenTargetCost = 0
+)
+
+func RegisterRoutes(r weave.Registry, auth x.Authenticator) {
+	r = migration.SchemaMigratingRegistry("username", r)
+
+	b := NewTokenBucket()
+	r.Handle(RegisterTokenMsg{}.Path(), &registerTokenHandler{auth: auth, bucket: b})
```
Comment (ethanfrey) [inline x/username/handlers.go]: I like use of. Path rather than reusing a constant. Cleaner and less error prone

File: cmd/bnsd/app/codec.proto [hunk 9]
```diff
@@ -131,6 +139,9 @@ message ProposalBatchMsg {
       gov.UpdateElectorateMsg update_electorate_msg = 77;
       gov.UpdateElectionRuleMsg update_election_rule_msg = 78;
       gov.TextResolutionMsg text_resolution_msg = 79;
+      username.RegisterTokenMsg register_token_msg = 80;
+      username.ChangeTokenOwnerMsg change_token_owner_msg = 81;
+      username.ChangeTokenTargetMsg change_token_target_msg = 82;
```
Comment (ruseinov) [inline cmd/bnsd/app/codec.proto]: I'm still not convinced that when we have potentially several targets for each nft it has to be done via batch, because we might hit a limitation here, especially when there are other transactions in the same batch. 

While it looks like a neat trick to simplify the NFT logic, I'd prefer to have slices in these messages. That allows for more api flexibility.

File: x/username/handlers.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	r.Handle(ChangeTokenOwnerMsg{}.Path(), &changeTokenOwnerHandler{auth: auth, bucket: b})
+	r.Handle(ChangeTokenTargetMsg{}.Path(), &changeTokenTargetHandler{auth: auth, bucket: b})
+}
+
+type registerTokenHandler struct {
+	auth   x.Authenticator
+	bucket orm.ModelBucket
+}
+
+func (h *registerTokenHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
```
Comment (ethanfrey) [inline x/username/handlers.go]: Very nice, clean, simple implementations of all these handlers. After this weekend, I was really looking for shirt contracts. This is nice example in any case. Even if we don't use this in Bnsd, it is a wonderful example for reference or tutorial

File: x/username/username.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/iov-one/weave/errors"
+)
+
+// Username represents a name registered on certain domain. A valid username is
+// in format <name>*<domain>. Username is case sensitive.
+type Username string
+
+// ParseUsername returns a valid username instance, extracted from given string
+// representation.
+func ParseUsername(s string) (Username, error) {
```
Comment (ethanfrey) [inline x/username/username.go]: I would store these two, name and domain, as two items in the db. And only combine in the ui or cli. Right now we have no domain support, but I could see an easy migration path with this implementation. Much easier to update and adapt in the future

**Comment by ruseinov** [comment]
Good work! 

Two main points I'd request to improve:
1. Message naming
2. The ability to update targets without batch. Ideally, since we want to point to multiple targets - the transaction should override the whole field in the model. This is actually one more important point to do it in one message and not batch.

Consider the scenario where you want to update targets field here:
```
message Token {
  weave.Metadata metadata = 1;
  // Target specifies where this username token points to.
  repeated Location targets = 2 [(gogoproto.nullable) = false];
```
What is our logic when you are updating this via batch with a message like this: 
```
message ChangeTokenTargetMsg {
  weave.Metadata metadata = 1;
  // Username is the unique name of the token, for example alice*iov
  string username = 2 [(gogoproto.casttype) = "Username"];
  // New target is a blockchain address location that this token should point
  // to after the change.
  Location new_target = 3 [(gogoproto.nullable) = false];
}
```

Do we override or just append to the slice? If we change this message to accept a set of new_targets we can simplify the logic to always override the previous value in the model.

File: spec/gogo/x/username/codec.proto (hunk 1 above)
Comment (husio) [reply spec/gogo/x/username/codec.proto]: Done in https://github.com/iov-one/weave/pull/771/commits/de3d772fd6c3508e8db76d2d169399f583a431f7 I have renamed it to `BlockchainAddress`.

File: x/username/codec.proto (hunk 3 above)
Comment (husio) [reply x/username/codec.proto]: I think being explicit is better. `ChangeTokenOwnerMsg` does not require an explanation of what it does. There is an `Token.Owner` field and with this message we can change it.

With `TransferTokenMsg` the name is slightly shorter, but you must read an explanation of what it does. I am not sure what is the adventage of this name.

File: x/username/codec.proto (hunk 4 above)
Comment (husio) [reply x/username/codec.proto]: Similar to https://github.com/iov-one/weave/pull/771#discussion_r294287539 I think a verbose name is in this case much better as it requires no explanation.

File: spec/gogo/x/username/codec.proto (hunk 2 above)
Comment (husio) [reply spec/gogo/x/username/codec.proto]: Implemented in https://github.com/iov-one/weave/pull/771/commits/288a34c4bafd8780ebedba1352d29310f477e28b

File: cmd/bnsd/app/codec.proto (hunk 9 above)
Comment (husio) [reply cmd/bnsd/app/codec.proto]: Is what was implemented in https://github.com/iov-one/weave/pull/771/commits/288a34c4bafd8780ebedba1352d29310f477e28b is enough or you had something else in mind?

File: x/username/codec.proto (hunk 7 above)
Comment (husio) [reply x/username/codec.proto]: Done in https://github.com/iov-one/weave/pull/771/commits/288a34c4bafd8780ebedba1352d29310f477e28b

Comment (husio) [reply x/username/codec.proto]: Done in https://github.com/iov-one/weave/pull/771/commits/288a34c4bafd8780ebedba1352d29310f477e28b

File: x/username/username.go (hunk 11 above)
Comment (husio) [reply x/username/username.go]: Relevant discussion https://github.com/iov-one/weave/issues/767#issuecomment-501987100

Do you have any use case in mind that storing those values separately makes easier? Right now, the `<name*domain>` string is used only as the token key. It is not stored as a value inside of the token. Do you want to store then additionally (in other words, add it to the model)?

**Comment by husio** [comment]
I have addressed all the comments (code change for all except messages naming).

I will work now on adding tests and removing `nft/username` from the `bnsd`.

File: x/username/handlers.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+	return &weave.CheckResult{GasAllocated: changeTokenTargetCost}, nil
+}
+
+func (h *changeTokenTargetsHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, token, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	token.Targets = msg.NewTargets
```
Comment (ethanfrey) [inline x/username/handlers.go]: The code looks good and addresses all my concerns.
The one issue I have is here. And this is a design question.

As it is currently written in (old) nft/username, it is a PATCH to add or remove names.
This defines it as a POST to overwrite them.

POST is much simpler to do on the server. But makes it a bit harder for the client to perform the action "add an ethereum address to my username".

Comment (ethanfrey) [reply x/username/handlers.go]: We can keep it as you implemented it. Just wanted to bring it up as a discussion point

File: x/username/username.go (hunk 11 above)
Comment (ethanfrey) [reply x/username/username.go]: Hmmm... fair enough. Leave it as is

File: x/username/username.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (u Username) Validate() error {
+	if !validUsername(string(u)) {
+		return errors.Wrap(errors.ErrInput, "invalid username")
+	}
+	return nil
+}
+
+var validUsername = regexp.MustCompile(`\w{3,}\*\w{3,}`).MatchString
```
Comment (ethanfrey) [inline x/username/username.go]: Current username requirements are here 
https://github.com/iov-one/weave/blob/master/cmd/bnsd/x/nft/username/msg.go#L20
 (it would be good to keep those rules for the part before the *)

File: x/username/model.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+func init() {
+	migration.MustRegister(1, &Token{}, migration.NoModification)
+}
+
+func (ba *BlockchainAddress) Validate() error {
+	switch n := len(ba.BlockchainID); {
+	case n < 3:
+		return errors.Wrap(errors.ErrInput, "blockchain ID too short")
```
Comment (ethanfrey) [inline x/username/model.go]: Nice to give special errors for length.
Can you also assert the values? To avoid emojis and other such nonsense

https://github.com/iov-one/weave/blob/master/cmd/bnsd/x/nft/username/model.go#L17
`var validBlockchainID = regexp.MustCompile(`^[a-zA-Z0-9_.-]{4,32}$`).Match`

File: x/username/model.go [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+	case n > 32:
+		return errors.Wrap(errors.ErrInput, "blockchain ID too long")
+	}
+	switch n := len(ba.Address); {
+	case n < 3:
+		return errors.Wrap(errors.ErrInput, "address too short")
+	case n > 1024:
```
Comment (ethanfrey) [inline x/username/model.go]: Woah, that is huuuuuge.
Can you trim it down to 128 bytes? I think 64 is fine, but let's give extra space.

Can you think of any usecase where we need more?

File: x/username/handlers.go (hunk 12 above)
Comment (husio) [reply x/username/handlers.go]: This is a great question. I always tend to do overwrite (POST) for two reasons.
1. It is a tiny bit easier to implement and much easier to explain how it works,
2. The client is always having a full state. Sending the full list should never be an issue. Sometimes it might be even easier than building the diff.

If this is an issue, we can always change it or introduce another message which provides different update strategy.

File: x/username/model.go (hunk 15 above)
Comment (husio) [reply x/username/model.go]: I have picked all validation rules without any knowledge what makes a good value. They are all to be refined and declared properly. 
128 bytes sounds great.

File: x/username/username.go (hunk 13 above)
Comment (husio) [reply x/username/username.go]: Addressed in https://github.com/iov-one/weave/pull/771/commits/65b0676f0ab27e05440c78d7ab7dda58f44c73cc

File: x/username/codec.proto (hunk 3 above)
Comment (alpe) [reply x/username/codec.proto]: IMHO `Transfer` is the business operation while `change token owner` is the technical implementation that happens under the hood. We should not expose implementation details in the API. 
Maybe it is even better to name it analogue to [cash.SendMsg](https://github.com/iov-one/weave/blob/master/x/cash/codec.proto#L21).

File: x/username/model.go (hunk 14 above)
Comment (husio) [reply x/username/model.go]: Addressed in https://github.com/iov-one/weave/pull/771/commits/9eef83f3027589acba690e2f500002722ef21cca

File: x/username/codec.proto (hunk 3 above)
Comment (husio) [reply x/username/codec.proto]: Makes sense.

File: x/username/handlers.go (hunk 12 above)
Comment (alpe) [reply x/username/handlers.go]: We should not push complexity to the client. A full overwrite is problematic in any WYSIWYG environment or where state gets updated quickly. I also want to highlight the asynchronous nature of our system. All the TXs get queues before they are applied and the order is not static as well.
A patch would make the system more resilient as we would not overwrite by adds and could fail on non existing removes.

Comment (ruseinov) [reply x/username/handlers.go]: Yeah, I think I'm guilty of suggesting the full overwrite too as it seemed easier to reason about in terms of server-side implementation, but indeed in blockchain case it might be wiser to do a patch.

**Comment by husio** [comment]
When including `x/username` [messages in the `bnsd` protobuf](https://github.com/iov-one/weave/blob/tiny_nft_username_issue_767/cmd/bnsd/app/codec.proto#L67-L69) is skipping the namespace prefix. This means that for example declaration 
```
username.RegisterTokenMsg register_token_msg = 80;
```
is exposed to non Go clients (ie Javascript) as `RegisterTokenMsg`. Message declarations are also pulled to a single file, without their prefix.

How about renaming username's `Token` message to `UsernameToken` or some shorter version like  `UToken`? This would cause all messages to contain the namespace in their name.
* `TransferTokenMsg` -> `TransferUsernameTokenMsg`
* `RegisterTokenMsg` -> `RegisterUTokenMsg`

---

This is similar to how it is done in [escrow](https://github.com/iov-one/weave/blob/tiny_nft_username_issue_767/cmd/bnsd/app/codec.proto#L40-L42) or [aswap](https://github.com/iov-one/weave/blob/tiny_nft_username_issue_767/cmd/bnsd/app/codec.proto#L57-L59) messages.
Because of naming, it must be confusing to use [currency](https://github.com/iov-one/weave/blob/tiny_nft_username_issue_767/cmd/bnsd/app/codec.proto#L46) or [cash](https://github.com/iov-one/weave/blob/tiny_nft_username_issue_767/cmd/bnsd/app/codec.proto#L38) messages.

**Comment by alpe** [comment]
Ja, good point about the naming. Let's rename the messages for the clients. I prefer `UsernameToken` over `UToken`. But even better would be just `Username` like TransferUsernameMsg` as the `Token` is an implementation detail for me.

**Comment by ruseinov** [comment]
Yeah, same here. I like the idea as we had some good patterns but didn't utilise the same one throughout our codebase.

About message names - I would prefer longer name rather than `UToken`, cause it does not ultimately matter for the codebase and it's nice for it to be self-explanatory.

File: x/username/handlers_test.go [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+						{BlockchainID: "bc_2", Address: []byte("addr2")},
+					},
+				},
+			},
+			Auth: &weavetest.Auth{Signer: bobbyCond},
+		},
+		"username must be unique": {
+			Tx: &weavetest.Tx{
+				Msg: &RegisterTokenMsg{
+					Username: "alice@iov",
```
Comment (alpe) [inline x/username/handlers_test.go]: q: how does this test pass? In your test setup (L86) you have registered `alice*iov` which is not equal to `alice@iov`

**Comment by alpe** [comment]
I had a chance to look at the code a bit more. Good work! 
I am wondering a bit about the tests passing with `@iov` in the data and `*iov` in the setup but I have not looked deeper into this.
To move this forward, I think we need to define some edge cases a bit better. Especially the empty target list. I tried to comment on the tests.
I would also suggest that we move this from `x/username` (framework level) to `cmd/bnsd/x` (app level) package

File: x/username/handlers_test.go [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+)
+
+func TestRegisterTokenHandler(t *testing.T) {
+	var (
+		aliceCond = weavetest.NewCondition()
+		bobbyCond = weavetest.NewCondition()
+	)
+
+	cases := map[string]struct {
+		Init           func(t testing.TB, db weave.KVStore)
```
Comment (alpe) [inline x/username/handlers_test.go]: req: not used

File: x/username/msg.go [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (m *RegisterTokenMsg) Validate() error {
+	if err := m.Metadata.Validate(); err != nil {
+		return errors.Wrap(err, "metadata")
+	}
+	if err := m.Username.Validate(); err != nil {
+		return errors.Wrap(err, "username")
+	}
+	if len(m.Targets) == 0 {
```
[the remark this answers is not in the document]
Comment (husio) [reply x/username/msg.go]: I don't know. @kmw101 should a user be able to register a username without an address? If yes, can a user delete all addresses from his username?

File: x/username/handlers_test.go [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+
+			b := NewTokenBucket()
+			_, err := b.Put(db, []byte("alice*iov"), &Token{
+				Metadata: &weave.Metadata{Schema: 1},
+				Targets: []BlockchainAddress{
+					{BlockchainID: "unichain", Address: []byte("756e69636f696e2d310a")},
+				},
+				Owner: aliceCond.Address(),
+			})
+			assert.Nil(t, err)
```
[the remark this answers is not in the document]
Comment (husio) [reply x/username/handlers_test.go]: It produces shitty output. I use it only in setup code where I never expect it to fail.

File: x/username/handlers_test.go (hunk 16 above)
Comment (husio) [reply x/username/handlers_test.go]: Tests were wrong. All is fixed now.

File: x/username/msg.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (kmw101) [reply x/username/msg.go]: For MvP, they can't. 

It was one of the many ideas floating around, if we did it then we would need to create the functionality for add/remove one or all names linked to a native address (and we could also get into the user selecting which chain to connect to and manage that process!). 

Let's keep this simple.

[the remark this answers is not in the document]
Comment (husio) [reply x/username/msg.go]: I have just implemented it :sob:

File: x/username/handlers_test.go [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+					Username: "does-not-exist@iov",
+					NewTargets: []BlockchainAddress{
+						{BlockchainID: "hydracoin", Address: []byte("68796472612d310a")},
+					},
+				},
+			},
+			WantCheckErr:   errors.ErrNotFound,
+			WantDeliverErr: errors.ErrNotFound,
+			Auth:           &weavetest.Auth{Signer: bobbyCond},
+		},
```
[the remark this answers is not in the document]
Comment (husio) [reply x/username/handlers_test.go]: Done in https://github.com/iov-one/weave/pull/771/commits/63338ad9db9b8e073154e26f1eff0a1ad1630441

File: x/username/handlers_test.go [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+	cases := map[string]struct {
+		Tx             weave.Tx
+		Auth           x.Authenticator
+		WantCheckErr   *errors.Error
+		WantDeliverErr *errors.Error
+	}{
+		"success": {
+			Tx: &weavetest.Tx{
+				Msg: &ChangeTokenOwnerMsg{
+					Username: "alice@iov",
```
[the remark this answers is not in the document]
Comment (husio) [reply x/username/handlers_test.go]: Done in https://github.com/iov-one/weave/pull/771/commits/156f6b194ee2a75ac2976b411ceb6ebc5c1fa29b

File: x/username/handlers.go [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+	var token Token
+	if err := h.bucket.One(db, msg.Username.Bytes(), &token); err != nil {
+		return nil, nil, errors.Wrap(err, "cannot get token from database")
+	}
+
+	if !h.auth.HasAddress(ctx, token.Owner) {
+		return nil, nil, errors.Wrap(errors.ErrUnauthorized, "only the token owner can execute this operation")
+	}
+
+	return &msg, &token, nil
```
[the remark this answers is not in the document]
Comment (husio) [reply x/username/handlers.go]: It is easier to not enforce that. Preventing from this does not bring any benefit.
Tests added in https://github.com/iov-one/weave/pull/771/commits/63338ad9db9b8e073154e26f1eff0a1ad1630441

File: x/username/handlers_test.go [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+			Tx: &weavetest.Tx{
+				Msg: &ChangeTokenOwnerMsg{
+					Username: "does-not-exist@iov",
+					NewOwner: bobbyCond.Address(),
+				},
+			},
+			WantCheckErr:   errors.ErrNotFound,
+			WantDeliverErr: errors.ErrNotFound,
+			Auth:           &weavetest.Auth{Signer: bobbyCond},
+		},
```
[the remark this answers is not in the document]
Comment (husio) [reply x/username/handlers_test.go]: Done in https://github.com/iov-one/weave/pull/771/commits/63338ad9db9b8e073154e26f1eff0a1ad1630441

**Comment by husio** [comment]
@alpe I have addressed all the comments. Please have a look again.

File: x/username/msg_test.go [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+		"invalid username": {
+			Msg: &ChangeTokenTargetsMsg{
+				Metadata: &weave.Metadata{Schema: 1},
+				Username: "xx",
+				NewTargets: []BlockchainAddress{
+					{BlockchainID: "blobchain", Address: []byte("1234567890")},
+				},
+			},
+			Want: errors.ErrInput,
+		},
```
[the remark this answers is not in the document]
Comment (husio) [reply x/username/msg_test.go]: Done in https://github.com/iov-one/weave/pull/771/commits/e8238a3da70c0614b162665a962705b6c5dbb139

File: cmd/bnsd/x/username/handlers.go [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+	return &weave.CheckResult{GasAllocated: changeUsernameTokenTargetCost}, nil
+}
+
+func (h *changeUsernameTokenTargetsHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, token, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	token.Targets = msg.NewTargets
```
Comment (ruseinov) [inline cmd/bnsd/x/username/handlers.go]: I remember us talking about not moving complexity to the frontend and using a Patch here. I haven't really followed the conversation after that, but I assume @alpe agreed with this approach or otherwise this change here is missing.

**Comment by ruseinov** [comment]
Briefly looked through the code, the implementation seems sane to me, one unresolved question is PATCH vs POST for new targets.

File: cmd/bnsd/x/username/handlers.go (hunk 25 above)
Comment (husio) [reply cmd/bnsd/x/username/handlers.go]: This is discussed in https://github.com/iov-one/weave/pull/771#discussion_r294515327

**Comment by alpe** [comment]
Thanks for the changes. let's resolve `PATCH vs POST`. Otherwise the code looks good 💐

File: x/username/handlers.go (hunk 12 above)
Comment (husio) [reply x/username/handlers.go]: Who is the person to decide how this should work? @kmw101 ?

File: cmd/bnsd/x/username/handlers.go (hunk 25 above)
Comment (ruseinov) [reply cmd/bnsd/x/username/handlers.go]: Yeah, so there was no further discussion. 

I'll be honest and say that to me the current implementation works just fine. The PATCH method does not really resolve issues with stuff being executed sequentially, because a PATCH can possibly contain all the same changes as POST and lead to similar issues, because the client will be using an outdated version of information stored in buckets.

So to me it boils down to complexity on the client vs server. And as I look at how this should properly be implemented on the client (and that's introspection of the current values to decide what the patch would be) - it looks like we'll have a similar amount of code either way.
Which leads me to conclusion that simplified logic on the backend won't hurt.

File: x/username/handlers.go (hunk 12 above)
Comment (husio) [reply x/username/handlers.go]: I have discussed this with @alpe and the `PATCH` solution would require the following rules

```
// ChangeTokenTargetsMsg is a request to change the address that this token
// points to. Only the owner of a token can request this operation.
message ChangeUsernameTokenTargetsMsg {
  weave.Metadata metadata = 1;
  // Username is the unique name of the token, for example alice*iov
  string username = 2 [(gogoproto.casttype) = "Username"];
  //
  // - if the .Address  is not empty then set/overwrite. 
  // - if the .Address is empty (nil) then delete 
  // - .Blokchain attribute value must be unique in the list.
  // - the new .Address value can be the same as the current value
  // - deleting an item that does not exist fails
  
  // In a different message (register)
  // - you cannot register a token with a list containing an empty .Blockchain value
  repeated BlockchainAddress diff = 3 [(gogoproto.nullable) = false];
}
```
The problem with this approach is that it is a very complex logic to implement. The current regexp + lenght check must be replaced with custom set of rules, different for each message.
The biggest issues are:

1. [`BlockchainAddress.Validate`](https://github.com/iov-one/weave/blob/e8238a3da70c0614b162665a962705b6c5dbb139/cmd/bnsd/x/username/model.go#L17-L27) can no longer validate the state as the validation rules differs depending on the message that is using it,
2. You need to have access to the database in order validate the `BlockchainAddress` collection,
3. It is easy to create an invalid configuration, not only by using an invalid data (ie invalid blockchain ID), but by providing a setup that leads to a logic violation

This is the last outstanding issue in this pull request. Should we proceed with the above implementation, try something else or merge the current state?

Comment (alpe) [reply x/username/handlers.go]: Post vs Patch is not blocking this PR IMHO. It would be nice to have it for the client but we should stick to our timeline. Patch can to into a new PR if we want to do it.

Comment (husio) [reply x/username/handlers.go]: Unless there will be any objections, I will merge this within an hour. :racehorse:
